src/lib/when.ts

```diff
@@ -0,0 +1,92 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {directive, reparentNodes, DirectiveFn, NodePart} from '../core.js';
+
+type PartCache = [NodePart, NodePart, DocumentFragment];
```

Comment:
Is there a reason to have this an Array instead of an Object with named keys?

```javascript
{
  ifPart: NodePart,
  elsePart: NodePart,
  container: DocumentFragment
}
```

Reply:
Agreed, but a small naming nit: since this isn't exactly an `if` statement, I'd name the properties semantically as `truePart` and `falsePart`

Reply:
I see what you mean. I started out with just two parts, so I made it a tuple to keep it simple and because I did the reverse() later on. Later on I added the container element.